Encryption.md

```diff
@@ -19,216 +19,544 @@
-might go unnoticed. 
+The file footer can be either encrypted or left as a plaintext. In an encrypted footer mode, 
+a new Thrift structure with file crypto metadata is added to the file. This metadata provides 
+information about the file encryption algorithm and the footer encryption key. 
 
-The `AesGcmV1` and `AesGcmCtrV1` structures contain an optional `aad_metadata` field that can 
-be used by a reader to retrieve the AAD string used for file encryption. The maximal allowed
-length of `aad_metadata` is 512 bytes.
+In a plaintext footer mode, the contents of the `FileMetaData` structure is visible, but it can 
+be optionally signed with a footer key in order to verify its integrity. The `FileMetaData` keeps
+the file encryption algorithm, the footer signing key description and a flag indicating whether 
+a footer signature is added to the file.
 
-Parquet-mr/-cpp implementations use the RBG-based IV construction as defined in the NIST 
-SP 800-38D document for the GCM ciphers (section 8.2.2).
+For encrypted columns, the following modules are always encrypted, with the same column key: 
+pages and  `PageHeader`'s (both dictionary and data), `ColumnIndex`'s, `OffsetIndex`'s.  If the 
+column key is different from the footer encryption key, the `ColumnMetaData` structure is serialized 
+separately and encrypted with the column key. In this case, the `ColumnMetaData` is also considered to be 
+a module.  
 
+There are two module types: data modules (pages) and Thrift modules (all Thrift structures that 
+are serialized separately).
 
-### AES_GCM_V1
-All modules are encrypted with the AES-GCM cipher. The authentication tags (16 bytes) are 
-written after each ciphertext. The IVs (12 bytes) are written before each ciphertext.
 
-### AES_GCM_CTR_V1
-Thrift modules are encrypted with the AES-GCM cipher, as described above. 
-The pages are encrypted with AES-CTR, where the IVs (16 bytes) are written before each 
-ciphertext.
+## 4. Encryption Algorithms and Keys
+Parquet encryption algorithms are based on the standard AES ciphers for symmetric encryption. 
+AES is supported in Intel and other CPUs with hardware acceleration of crypto operations 
+(“AES-NI”) - that can be leveraged, for example, by Java programs (automatically via HotSpot), 
+or C++ programs (via EVP-* functions in OpenSSL). Parquet supports all standard AES key sizes: 
+128, 192 and 256 bits. 
 
+Initially, two algorithms have been implemented, one based on a GCM mode of AES, and the 
+other on a combination of GCM and CTR modes.
 
-## File Format
+AES GCM is an authenticated encryption. Besides the data confidentiality (encryption), it 
+supports two levels of integrity verification (authentication): of the data (default), 
+and of the data combined with an optional AAD (“additional authenticated data”). The 
+default authentication allows to make sure the data has not been tampered with. An AAD 
```

Comment:
I think this should note that without an AAD, GCM validates that the encrypted bytes of a single encrypted buffer haven't been tampered with. I think it should also be clear that the "default" is when an AAD is not used.